### PR TITLE
userRoutes

### LINKED ISSUE #10: Create User Model

### LINKED ISSUE #11: Create User Controller

### LINKED ISSUE #12: Create User Routes

### LINKED ISSUE #15: Create Service Provider Controller

### LINKED ISSUE #16: Create Service Provider Routes

### PR DESCRIPTION
**Pull request by ScottBeverly**

### CODE REVIEW

File: Controllers/providerController.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+const getProviderByName = (req, res) => {
+    Provider.findOne({"name": req.params.name}).then(provider => {
+      res.status(200).json({
+          status: "ok",
+          provider
+        })  
+    })
+};
+
```
Comment (bcantello) [inline Controllers/providerController.js:26]: Add a getProviderById function. Id is more unique than name so I imagine we will want to use that whenever possible.

File: Controllers/providerController.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+const Provider = require('../Models/providerModel');
+
+const getAllProviders = (req, res) =>{
```
Comment (bcantello) [inline Controllers/providerController.js]: nit: add space after end of fat arrow. Some of your functions have a space, some don't. Should be consistent. Let's do spaces for all.

File: Controllers/providerController.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+const getProviderByName = (req, res) => {
+    Provider.findOne({"name": req.params.name}).then(provider => {
+      res.status(200).json({
+          status: "ok",
+          provider
+        })  
+    })
+};
+
+const createProvider = (req, res) =>{
```
Comment (bcantello) [inline Controllers/providerController.js]: nit: space after fat arrow

File: Controllers/providerController.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    const body = req.body
+    Provider.create(body).then(provider => {
+        Provider.find({}).then((providers) => {
+            res.status(200).json({
+                status: "ok",
+                msg: `Updated ${provider.id}: ${provider.name}`,
+                providers
+            })
+        })
```
Comment (bcantello) [inline Controllers/providerController.js]: Is this all necessary? Looks like the create function got blended with an update function?

File: Controllers/providerController.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    })
+};
+
+const updateProvider = (req, res) =>{
```
Comment (bcantello) [inline Controllers/providerController.js]: nit: white space after arrow

File: Controllers/providerController.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    const body = req.body
+    Provider.findOneAndUpdate({_id: req.params.id}, body).then((provider =>{
```
Comment (bcantello) [inline Controllers/providerController.js]: nit: white space after arrow

File: Controllers/providerController.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+module.exports = {
+    getAllProviders,
+    getProviderByName,
+    createProvider,
+    updateProvider,
+    deleteProvider
+};
+
+
```
Comment (bcantello) [inline Controllers/providerController.js:73]: nit: remove extra white space

File: Controllers/userController.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
+const User = require('../Models/userModel');
+
+const getAllUsers = (req, res) =>{
```
Comment (bcantello) [inline Controllers/userController.js]: nit: white space after fat arrow

File: Controllers/userController.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+const getUsersByName = (req, res) => {
+    User.findOne({"FirstName": req.params.Name}).then(user => {
+      res.status(200).json({
+          status: "ok",
+          user
+        })  
+    })
+};
+
```
Comment (bcantello) [inline Controllers/userController.js:26]: Add getUserById function

File: Controllers/userController.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+const createUser = (req, res) =>{
+    const body = req.body
+    User.create(body).then(user => {
+        User.find({}).then((users) => {
+            res.status(200).json({
+                status: "ok",
+                msg: `Updated ${user.id}: ${user.name}`,
+                users
+            })
```
Comment (bcantello) [inline Controllers/userController.js]: Is this necessary for a create function?

File: Controllers/userController.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+const getUsersByName = (req, res) => {
+    User.findOne({"FirstName": req.params.Name}).then(user => {
+      res.status(200).json({
+          status: "ok",
+          user
+        })  
+    })
+};
+
+const createUser = (req, res) =>{
```
Comment (bcantello) [inline Controllers/userController.js]: nit: space after arrow

File: Controllers/userController.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            res.status(200).json({
+                status: "ok",
+                msg: `Updated ${user.id}: ${user.name}`,
+                users
+            })
+        })
+    })
+};
+
+const updateUser = (req, res) =>{
```
Comment (bcantello) [inline Controllers/userController.js]: nit: space after arrow

File: Controllers/userController.js [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    const body = req.body
+    User.findOneAndUpdate({_id: req.params.id}, body).then((user =>{
```
Comment (bcantello) [inline Controllers/userController.js]: nit: white space after arrow

File: Controllers/userController.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+            res.status(200).json({
+                status: "ok",
+                msg: `Updated ${user.id}: ${user.name}`,
+                users
+            })
+        })
+    }))
+};
+
+const deleteUser = (req, res) =>{
```
Comment (bcantello) [inline Controllers/userController.js:52]: nit: space after arrow

File: Controllers/userController.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    User.findOneAndDelete({_id: req.params.id}).then((user) =>{
```
Comment (bcantello) [inline Controllers/userController.js]: nit space after arrow

File: Controllers/userController.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+module.exports = {
+    getAllUsers,
+    getUsersByName,
+    createUser,
+    updateUser,
+    deleteUser
+};
+
+
```
Comment (bcantello) [inline Controllers/userController.js:73]: nit: remove extra white space

File: Models/userModel.js [hunk 17]
```diff
@@ -0,0 +1,15 @@
+const mongoose = require('../DataBase/connection');
+const UserSchema = mongoose.Schema({
+
```
Comment (bcantello) [inline Models/userModel.js:3]: nit: remove extra white space

File: Models/userModel.js [hunk 18]
```diff
@@ -0,0 +1,15 @@
+const mongoose = require('../DataBase/connection');
```
Comment (bcantello) [inline Models/userModel.js:1]: nit: Add white space between dependencies and functions

File: Models/userModel.js [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+const UserSchema = mongoose.Schema({
+   
+    name: String,
+    zip: Number,
```
Comment (bcantello) [inline Models/userModel.js]: We should probably include street address for the user. Provider needs to get to their house

File: Models/userModel.js [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+    pet: [
+        {
+            petName: String,
+            age: Number,
+            breed: String
+        }
+    ]
+})
```
Comment (bcantello) [inline Models/userModel.js:20]: nit: space between end of function and exports

File: Models/userModel.js [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+const User = mongoose.model("User", UserSchema);
+module.exports = User;
```
Comment (bcantello) [inline Models/userModel.js:22]: nit: add whitespace to end of file

File: Routes/providerRoutes.js [hunk 22]
```diff
@@ -0,0 +1,17 @@
+const express = require("express");
+const router = express.Router();
+
```
Comment (bcantello) [inline Routes/providerRoutes.js]: nit: remove white space. const controller is a dependency declaration

File: Routes/providerRoutes.js [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+const controller = require('../Controllers/providerController');
+
+router.get('/', controller.getAllProviders);
+
+router.get('/:name', controller.getProviderByName);
+
```
Comment (bcantello) [inline Routes/providerRoutes.js:10]: add get provider by Id route after created

File: Routes/userRoutes.js [hunk 24]
```diff
@@ -0,0 +1,16 @@
+const express = require("express");
+const router = express.Router();
+
```
Comment (bcantello) [inline Routes/userRoutes.js]: remove white space

File: Routes/userRoutes.js [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+const controller = require('../Controllers/userController');
+
+router.get('/', controller.getAllUsers);
+
+router.get('/:name', controller.getUsersByName);
+
```
Comment (bcantello) [inline Routes/userRoutes.js:10]: add get by id route

File: Routes/userRoutes.js [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+router.post('/', controller.createUser);
+
+router.put('/:id', controller.updateUser);
+
+router.delete('/:id', controller.deleteUser);
+
+module.exports = router;
```
Comment (bcantello) [inline Routes/userRoutes.js:17]: nit: add white space at end of file

**Review by bcantello — CHANGES_REQUESTED** [review]

File: Models/userModel.js [hunk 27]
```diff
@@ -0,0 +1,15 @@
+const mongoose = require('../DataBase/connection');
+const UserSchema = mongoose.Schema({
```
Comment (bcantello) [inline Models/userModel.js:2]: Need to either instantiate a new instance of mongoose.Schema, or declare a global Schema variable.
- instantiate:  const UserSchema = new mongoose.Schema({
- declare:  const Schema = mongoose.Schema;
                   const UserSchema = new Schema({

**Review by bcantello — CHANGES_REQUESTED** [review]

File: Controllers/userController.js [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+const getAllUsers = (req, res) =>{
+    User.find({}).then(user => {
+        res.status(200).json({
+            status: "ok",
+            user
+        })
+    })
+};
+
```
Comment (bcantello) [inline Controllers/userController.js:11]: Add getUserAppointmentsByID function
We need to be able to pull all of the appointments that are relevant to a specific user.  something like:
```suggestion

const getAuthorsWithCookbooks = (req, res) => {
     Author.find({ _id: req.params.id })
     .populate("cookbooks")
     .then((authors) => {
         res.json(authors);
     });
};


```

File: Models/userModel.js (hunk 19 above)
Comment (bcantello) [reply Models/userModel.js]: Also should include a key value pair for email, and one for password as well

File: Models/userModel.js [hunk 29]
```diff
@@ -0,0 +1,15 @@
...
+   
+    name: String,
+    zip: Number,
+    pet: [
+        {
+            petName: String,
+            age: Number,
+            breed: String
+        }
+    ]
```
Comment (bcantello) [inline Models/userModel.js]: I think we probably need a collection for pets, with its own schema and controllers for creating, updating, and deleting pets.

Comment (bcantello) [reply Models/userModel.js]: I created the Pet model, controller, and router. The pets section of the user schema should be updated to the following, and a preferredProvider section should be added as well:
```suggestion
    pet: [
        {
            ref: "Pet",
            type: mongoose.Schema.Types.ObjectId
        }
    ],
    preferredProvider: [
        {
            ref: "Provider",
            type: mongoose.Schema.Types.ObjectId
        }
    ]
```

**Review by GracieEveritt — APPROVED** [review]

**Pull request merged**